blocks/monitoring/__init__.py

```diff
@@ -0,0 +1,193 @@
+    All constructor parameters are accessible as attributes.
+
+    """
+    def __init__(self, aggregation_scheme, initialization_updates=None,
+                 accumulation_updates=None, readout_expression=None, **kwargs):
+        super(VariableMonitorChannel, self).__init__(**kwargs)
+        if initialization_updates is None:
+            initialization_updates = []
+        if accumulation_updates is None:
+            accumulation_updates = []
+        update_instance(self, locals())
+
+
+class FunctionMonitorChannel(MonitorChannel):
+    """A function whose output should be monitored
+
+    Parameters
+    ----------
+    function : callable
+        A callable function, which takes a model, training dataset, and set
+        of monitoring datasets as arguments.
+
+    Notes
+    -----
+    The values returned by callable monitor channels can be any Python
+    object.
+
+    """
+    def __init__(self, function, **kwargs):
+        super(FunctionMonitorChannel, self).__init__(**kwargs)
+        assert callable(function)
+        update_instance(self, locals())
+
+
+class AggregationScheme(object):
+    """Specify how to incrementally evaluate a Theano variable on data.
+
+    An AggregationScheme allocates :class:`VariableMonitorChannel`s that
+    can incrementally compute the value of a Theano variable on a full
+    dataset by aggregating partial results computed on multiple batches.
+
+    The AggregationScheme should be attached via the tag
+    `aggregation_scheme` to a Theano variable which computes the desired
+    value on a single batch.
+
+    Parameters
+    ----------
+    expression: Theano variable
+        expression that computes the desired value on a single batch.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def get_channel(self):
+        """Return a new VariableMonitorChannel for this variable."""
+        pass
+
+
+class Mean(AggregationScheme):
+    """Aggregation scheme which computes the mean.
+
+    Parameters
+    ----------
+    numerator : Theano variable
+        The expression for the numerator e.g. the likelihood of a
+        minibatch.
+    denominator : Theano variable
+        The expression for the denominator e.g. the size of the batch.
+
+    """
+    def __init__(self, numerator, denominator):
+        update_instance(self, locals())
+
+    def get_channel(self):
+        numerator_accumulated = shared_expression(self.numerator)
+        denominator_accumulated = shared_expression(self.denominator)
+        initialization_updates = [(numerator_accumulated, 0.0),
+                                  (denominator_accumulated, 0.0)]
+        accumulation_updates = [(numerator_accumulated,
+                                 numerator_accumulated + self.numerator),
+                                (denominator_accumulated,
+                                 denominator_accumulated + self.denominator)]
+        return VariableMonitorChannel(
+            aggregation_scheme=self,
```

Comment:
Here you should pass a name. I didn't do it, because the name was the expression's name.